runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/GrpcDirectGetWorkStream.java

```diff
@@ -316,5 +315,20 @@ private void runAndReset() {
       workTimingInfosTracker.reset();
       data = ByteString.EMPTY;
     }
+
+    private Work.ProcessingContext.WithProcessWorkFn createProcessingContext(String computationId) {
+      return Work.createProcessingContext(computationId, getDataStream.get()::requestKeyedData)
+          .setWorkCommitter(workCommitter.get()::commit)
+          .setGetDataStream(getDataStream.get());
+    }
+
+    private Work.Watermarks createWatermarks(WorkItem workItem) {
```

Comment:
nit: how about making this static and passing in metadata? it makes it clearer when reading above that it uses metadatda for the watermarks not just hte work item.

can also have a single check that metadata is set in runAndReset

Reply:
done